ot/lowrank.py

```diff
@@ -0,0 +1,502 @@
+        Initialization strategy for gamma. 'rescale', or 'theory'
+    numItermax : int, optional. Default is 2000.
+        Max number of iterations for the Dykstra algorithm
+    stopThr : float, optional. Default is 1e-7.
+        Stop threshold on error (>0) in Dykstra
+    warn : bool, optional
+        if True, raises a warning if the algorithm doesn't convergence.
+    log : bool, optional
+        record log if True
+
+
+    Returns
+    ---------
+    Q : array-like, shape (n_samples_a, r)
+        First low-rank matrix decomposition of the OT plan
+    R: array-like, shape (n_samples_b, r)
+        Second low-rank matrix decomposition of the OT plan
+    g : array-like, shape (r, )
+        Weight vector for the low-rank decomposition of the OT
+    log : dict (lazy_plan, value and value_linear)
+        log dictionary return only if log==True in parameters
+
+
+    References
+    ----------
+    .. [65] Scetbon, M., Cuturi, M., & Peyré, G. (2021).
+        "Low-rank Sinkhorn Factorization". In International Conference on Machine Learning.
+
+    """
+
+    # POT backend
+    nx = get_backend(X_s, X_t)
+    ns, nt = X_s.shape[0], X_t.shape[0]
+
+    # Initialize weights a, b
+    if a is None:
+        a = unif(ns, type_as=X_s)
+    if b is None:
+        b = unif(nt, type_as=X_t)
+
+    # Compute rank (see Section 3.1, def 1)
+    r = rank
+    if rank is None:
+        r = min(ns, nt)
+    else:
+        r = min(ns, nt, rank)
+
+    if r <= 0:
+        raise ValueError("The rank parameter cannot have a negative value")
+
+    # Compute alpha
+    if alpha is None:
+        alpha = 1e-10
+
+    # Dykstra won't converge if 1/rank < alpha (see Section 3.2)
+    if 1 / r < alpha:
+        raise ValueError("alpha ({a}) should be smaller than 1/rank ({r}) for the Dykstra algorithm to converge.".format(
+            a=alpha, r=1 / rank))
+
+    # Low rank decomposition of the sqeuclidean cost matrix
+    M1, M2 = compute_lr_sqeuclidean_matrix(X_s, X_t, rescale_cost, nx)
+
+    # Initialize the low rank matrices Q, R, g
+    if init is None:
+        init = "random"
+
+    Q, R, g = _init_lr_sinkhorn(X_s, X_t, a, b, r, init, reg_init, seed_init, nx=nx)
+
+    # Gamma initialization
+    if gamma_init is None:
+        gamma_init = "rescale"
+
+    if gamma_init == "theory":
+        L = nx.sqrt(
+            3 * (2 / (alpha**4)) * ((nx.norm(M1) * nx.norm(M2)) ** 2) +
+            (reg + (2 / (alpha**3)) * (nx.norm(M1) * nx.norm(M2))) ** 2
+        )
+        gamma = 1 / (2 * L)
+
+    if gamma_init not in ["rescale", "theory"]:
+        raise (NotImplementedError('Not implemented gamma_init="{}"'.format(gamma_init)))
+
+    # -------------------------- Low rank algorithm ------------------------------
+    # see "Section 3.3, Algorithm 3 LOT"
+
+    for ii in range(100):
+        # Compute C*R dot using the lr decomposition of C
+        CR = nx.dot(M2.T, R)
+        CR_ = nx.dot(M1, CR)
+        diag_g = (1 / g)[None, :]
```

Comment:
I added in `lowrank_sinkhorn` : `Q = Q + 1e-16` and `R = R + 1e-16` to prevent this issue for Q and R.
I could to do same for `g` by adding `g = g + 1e-16`.